### PR TITLE
Decouple OAuth client authentication 

### PR DESCRIPTION
**Pull request by hasinthaindrajee**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

OAuth client authentication is decoupled from oauth endpoint logic. Basic Client Authenticator is included as the default authenticator.
-

### When should this PR be merged
ASAP
[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions
When product-is uses this changes, there should be another PR merged to product-is
[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+            } catch (IdentityOAuth2Exception e) {
+                throw new OAuthClientAuthnException(OAuth2ErrorCodes.INVALID_CLIENT, "Error while authenticating " +
+                        "client", e);
+            }
+        }
+    }
+
+    private void validateAuthenticationInfo(HttpServletRequest request, Map<String, List> contentMap)
+            throws OAuthClientAuthnException {
+        if (isAuthorizationHeaderExists(request)) {
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticator.java]: Aren't we follow the convention to put a line space between the method declaration and the first line of the method? This is relevant to some other places as well in this pr.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+            oAuthClientAuthnContext) throws OAuthClientAuthnException {
+
+        if (authorizationHeader == null) {
+            throw new OAuthClientAuthnException("Authorization header value is null");
+        }
+        String[] splitValues = authorizationHeader.trim().split(" ");
+        if (splitValues.length == 2) {
+            byte[] decodedBytes = Base64Utils.decode(splitValues[1].trim());
+            String userNamePassword = new String(decodedBytes, Charsets.UTF_8);
+            String[] credentials = userNamePassword.split(":");
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticator.java]: Shall we use constant for ":"

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuthRevocationRequestDTO.java [hunk 3]
```diff
@@ -79,4 +83,12 @@ public void setTokenType(String tokenType) {
...
     }
+
+    public OAuthClientAuthnContext getoAuthClientAuthnContext() {
+        return oAuthClientAuthnContext;
+    }
+
+    public void setoAuthClientAuthnContext(OAuthClientAuthnContext oAuthClientAuthnContext) {
+        this.oAuthClientAuthnContext = oAuthClientAuthnContext;
+    }
 }
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuthRevocationRequestDTO.java]: Please add a new line in the EOF.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    }
+
+    /**
+     * Get the name of the OAuth2 client authenticator.
+     *
+     * @return The name of the OAuth2 client authenticator.
+     */
+    @Override
+    public String getName() {
+        return "BasicOAuthClientCredAuthenticator";
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticator.java]: Add a new line

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticator.java [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+     * @return An array which has client id as the first element and secret as the second element.
+     * @throws OAuthClientAuthnException
+     */
+    protected static String[] extractCredentialsFromAuthzHeader(String authorizationHeader, OAuthClientAuthnContext
+            oAuthClientAuthnContext) throws OAuthClientAuthnException {
+
+        if (authorizationHeader == null) {
+            throw new OAuthClientAuthnException("Authorization header value is null");
+        }
+        String[] splitValues = authorizationHeader.trim().split(" ");
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticator.java]: use a constant.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticator.java [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        if (splitValues.length == 2) {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticator.java]: use a constant

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthnService.java [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        }
+    }
+
+    /**
+     * Fails authentication if multiple authenticators are eligible of handling the request.
+     *
+     * @param oAuthClientAuthnContext OAuth client authentication context.
+     */
+    private void failOnMultipleAuthenticators(OAuthClientAuthnContext oAuthClientAuthnContext) {
+        if (oAuthClientAuthnContext.getExecutedAuthenticators().size() > 1) {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthnService.java]: add a new line

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/CarbonOAuthTokenRequest.java [hunk 8]
```diff
@@ -192,4 +194,24 @@ public String getPkceCodeVerifier() {
...
+    public OAuthClientAuthnContext getoAuthClientAuthnContext() {
+        return oAuthClientAuthnContext;
+    }
+
+    public void setoAuthClientAuthnContext(OAuthClientAuthnContext oAuthClientAuthnContext) {
+        this.oAuthClientAuthnContext = oAuthClientAuthnContext;
+    }
+
+    private void setClientAuthnContext(HttpServletRequest request) {
+        Object oauthClientAuthnContextObj = request.getAttribute(OAuthConstants.CLIENT_AUTHN_CONTEXT);
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/model/CarbonOAuthTokenRequest.java]: add a new line

**Pull request merged**